File: src/particles.c
#define _USE_MATH_DEFINES
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <emscripten.h>

// 
// emcc -O3 particles.c -o wasm.js -s NO_EXIT_RUNTIME=1 -s EXPORTED_RUNTIME_METHODS=[ccall]

typedef struct {
    double x;           
    double y;           
    int isMoving;               
} Particle;

Particle *particles;
int n;
int width;
int height;
int length;
int r;

EM_JS(void, draw_particle, 
    (double x, double y, double r, int isMoving), {
        drawParticle(x, y, r, isMoving);
    }
)

double randomNum() {
    return (double)rand() / (double)RAND_MAX;
}

EMSCRIPTEN_KEEPALIVE
void initParticles(int amount, double radius) {
    n = amount;
    r = radius;
    particles = malloc(sizeof(Particle) * n);
    for (int i=0; i < n; i++) {
        particles[i].x = randomNum() * width;
        particles[i].y = randomNum() * height;
        particles[i].isMoving = 1;
    }
}

EMSCRIPTEN_KEEPALIVE
void reInitParticles(int amount, double radius) {
    n = amount;
    r = radius;
    particles = realloc(particles, sizeof(Particle) * n);
    for (int i=0; i < n; i++) {
        particles[i].x = randomNum() * width;
        particles[i].y = randomNum() * height;
        particles[i].isMoving = 1;

    }
}

EMSCRIPTEN_KEEPALIVE
void freeMemory() {
    free(particles);
}

void doCollision(Particle* p, int i) {
    for (int j = 0; j < n; j++) {
        if (i == j) continue;  
        if ( sqrt(pow(p->x - particles[j].x, 2) + pow(p->y - particles[j].y, 2)) <= 2*r ) {
            //if (!particles[i].isMoving) 
            p->isMoving = 0;
       }
    }
}

EMSCRIPTEN_KEEPALIVE
void releaseAll_wasm() {
    for (int i = 0; i < n; i++) particles[i].isMoving = 1;
}

void randomMove(Particle* p) {
    double v = randomNum() * 2 * M_PI;
    p->x += length * cos(v);
    p->y += length * sin(v);
}

void updateParticle(Particle* p, int i) {
    // move
    if (p->isMoving) randomMove(p);
    
    // wall collision
    if (p->x >= width) {
        p->x = width - r;
        p->isMoving = 0;
    } else if (p->x <= r) {
        p->x = r;
        p->isMoving = 0;
    }  
    if (p->y >= height) {
        p->y = height - r;
        p->isMoving = 0;
    } else if (p->y <= r) {
        p->y = r;
        p->isMoving = 0;
    }

    // particle collision
    doCollision(p, i);

    // draw particle
    draw_particle(p->x, p->y, r, p->isMoving);
}

EMSCRIPTEN_KEEPALIVE
void update() {
    for (int i = 0; i < n; i++) {
        updateParticle(&particles[i], i);
    }
}

EMSCRIPTEN_KEEPALIVE
void drawParticle(double x, double y, double r, int isMoving) {
    draw_particle(x, y, r, isMoving);
}

EMSCRIPTEN_KEEPALIVE
void setVariables(int w, int h, int l) {
    width = w;
    height = h;
    length = l;
}


EMSCRIPTEN_KEEPALIVE
int main() {
    return 0;
}

File: src/particles.js
const canvas = document.getElementById("particles");
const ctx = canvas.getContext('2d');
canvas.width = window.innerWidth;
canvas.height = window.innerHeight;
var anim;

var lengthSlider = document.getElementById("length");

var particles = [];
let amount = 100;
let r = 1.5;              // radius of particles
let l = 1;            // length that a particle moves between each frame
let color = "#00ff3c";       // color of particle
let stuckColor = "#6d46eb";  // color of a stuck particle

var fps = document.getElementById("fps");
var startTime = performance.now();
var frame = 0;

document.addEventListener('keydown', (e) => {
    if (e.key === ' ' || e.key === 'Spacebar') {
        if (document.getElementById('wasm').checked) {
            Module.ccall(
                'releaseAll_wasm',
                null,
                null,
                null
            );
        } else {
            releaseAll();
        }
    }
})

lengthSlider.oninput = () => {
    l = lengthSlider.value;
    Module.ccall(
        'setVariables',
        null,
        ['number', 'number', 'number', 'number'],
        [canvas.width, canvas.height, l, r]
    );
}

function updateParticles() {

    cancelAnimationFrame(anim);
    resetParticlesArrays();
    var value = document.getElementById("numOfParticles");
    amount = parseInt(value.options[value.selectedIndex].value);
    
    if (document.getElementById('wasm').checked) {
        console.log("wasm checked!");
        Module.ccall(
            'setVariables',
            null,
            ['number', 'number', 'number', 'number'],
            [canvas.width, canvas.height, l, r]
        );
        Module.ccall(
            'initParticles',
            null,
            ['number'],
            [amount]
        );
        animateParticles_wasm();
    } else {
            initParticles();  
            animateParticles();
        }
    }

function initParticles() {
    particles = []
    for (var i = 0; i < amount; i++) {
        let x = Math.random() * canvas.width;
        let y = Math.random() * canvas.height;
        particles.push(new Particle(x, y, true));
    }
}

class Particle {
    constructor(x, y, isMoving) {
        this.x = x;
        this.y = y;
        this.isMoving = isMoving;
    }

    draw() {
        ctx.beginPath();
        ctx.arc(this.x, this.y, r, 0, 2 * Math.PI, false);
        ctx.closePath();
        ctx.fillStyle = this.isMoving ? color : stuckColor;
        ctx.fill()
    }

    update(i) {
        if (this.isMoving) this.randomMove(l);
        
        if (this.x >= canvas.width-r) {
            this.x = canvas.width-r;
            this.isMoving = false;
        }
        else if (this.x <= r) {
            this.x = r;
            this.isMoving = false;
        }
        else if (this.y >= canvas.height-r) {
            this.y = canvas.height-r;
            this.isMoving = false;
        }
        else if (this.y <= r) {
            this.y = r;
            this.isMoving = false;
        }

        this.collision(i);

        this.draw();
    }

    collision(i) {
        for (var j = 0; j < amount; j++) {
            if (i === j) continue;
            if (Math.sqrt(Math.pow(this.x - particles[j].x, 2) + Math.pow(this.y - particles[j].y, 2)) <= 2*r) {
                this.isMoving = false;
            }
        }
    }

    randomMove(L) {
        let v = Math.random() * 2*Math.PI;
        this.x += L * Math.cos(v);
        this.y += L * Math.sin(v);
    }
   
    release() {
        this.isMoving = true;
    }

}

function releaseAll() {
    particles.forEach(p => p.release());
}

function animateParticles(time) {
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    frame++;
    if (time - startTime > 1000) {
        fps.innerHTML = (frame / ((time - startTime) / 1000)).toFixed(1);
        startTime = time;
        frame = 0;
      }
    for (let i = 0; i < particles.length; i++) {
        particles[i].update(i);
    }

    anim = requestAnimationFrame(animateParticles);
}

function animateParticles_wasm(time) {
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    //var startUpdateTime = performance.now();
    //var time = performance.now();
    frame++;
    if (time - startTime > 1000) {
        fps.innerHTML = (frame / ((time - startTime) / 1000)).toFixed(1);
        startTime = time;
        frame = 0;
    }
    Module.ccall(
        'update',
        null,
        null,
        null
    );
    //var endUpdateTime = performance.now();
    //console.log(`Updating and drawing all particles (ccall) took: ${endUpdateTime-startUpdateTime}`);
    anim = requestAnimationFrame(animateParticles_wasm);
}

// invoked from particles.c
function drawParticle(x, y, radius, isMoving) {
    ctx.beginPath();
    ctx.arc(x, y, radius, 0, 2 * Math.PI, false);
    ctx.closePath();
    ctx.fillStyle = isMoving ? color : stuckColor;
    ctx.fill()
}

function resetParticlesArrays() {
    particles = [];
    Module.ccall(
        'freeMemory',
        null,
        null,
        null
    )
}

// invoke C initParticles function
Module.onRuntimeInitialized = () => {
    // set w, h, l, r
    Module.ccall(
        'setVariables',
        null,
        ['number', 'number', 'number', 'number'],
        [canvas.width, canvas.height, l, r]
    );
    // init particles
    Module.ccall(
        'initParticles',
        null,
        ['number'],
        [amount]
    );
    initParticles();
    //animateParticles_wasm();
    animateParticles();
}